proposals/0000-asset-primitive.md

```diff
@@ -0,0 +1,233 @@
+- Can reasonably extend to support image optimizations in the future.
+- Works with SSR, where all assets must be known at build-time, with 100% accuracy.
+- Works across frameworks (`.jsx`, `.svelte`, `.vue`). Not an `.astro`-only solution.
+- Introduces an acceptable amount of boilerplate (as little as possible, or none).
+- Can power custom 3rd-party integrations and components.
+- Can power future, higher-level Astro core features.
+
+# Detailed design
+
+## The Asset Primitive
+
+```js
+import imageBook from '../assets/book.png';
+  // imageBook is type: ImageAsset (.png|.jpg|.jpeg|.webp)
+  // {url, width, height, blurHashUrl}
+import iconCookie from '../assets/cookie.svg';
+  // iconCookieis type: SvgAsset (.svg)
+  // {width, height, content}
+import robotoFont400 from '../assets/fonts/roboto-slab-v22-latin-regular.woff2';
+  // robotoFont400 is type: string
+  // [If no custom asset is defined, Astro will default to including the asset
+  // in your build and returning a single URL as the result. This is Vite default behavior.]
+```
+
+- The asset primitive is imported via ESM.
+- Astro will define two custom `Asset` types: `ImageAsset` and `SvgAsset`
+- All other assets will just return a URL path string to the asset, and include that asset in the final build.
+- Works in `.astro` and all component files (`.jsx`, `.svelte`, `.vue`, etc)
+- Builds on top of the accepted [`local:src`](https://github.com/withastro/rfcs/blob/assets-rfc/proposals/0011-relative-url-scheme.md) RFC.
+
+## The Asset Components
+
+```astro
+---
+import imageBook from '../assets/book.png';
+import iconCookie from '../assets/cookie.svg';
+import robotoFont400 from '../assets/fonts/roboto-slab-v22-latin-regular.woff2';
+import {Image, Icon, Font} from 'astro/components';
+---
+<html>
+<head>
+  <Font use={robotoFont400} family="Roboto Slab" weight={400} style="normal" />
+</head>
+<body>
+  <Icon use={IconCookie} />
+  <Image use={imageBook} />
+</body>
+</html>
+```
+
+- Astro will ship `Font`, `Icon`, and `Image` components that can take these asset types, and properly use them on the page:
+  - `Font`: Creates a `<style>` tag with a single `@font-family` definition
+  - `Image`: Creates a `<img>` tag with `src=` the asset URL
+  - `Icon`: renders the given `svg` content to HTML.
+- In the future, these can be extended to add new features. For example, the Image component can be extended to support image optimizations and automatic [blurhash](https://blurha.sh/) support.
+- Renders could also provide `Font`, `Icon` and `Image` components that are specific to each framework. This can be fleshed out in more detail if it is a concern, but it is strongly believed that this is possible and fairly trivial.
+
+## Other Notes
+
+- This is trivial to implement in Astro today. Vite provides hooks for these sorts of imports, so we would just need to create two Vite plugins inside of Astro: one for SVGs and one for images.
```

Comment:
What would these plugins do, specifically? I'm assuming the SVG one returns SVG as raw. If that's the case I think it should be explicit in the proposal. It's unclear to me what the image plugin is intended to do.

Reply:
The data type is highlighted above, but I'll add a note here to make it more clear